cpp/src/arrow/compute/kernels/scalar_string_test.cc

```diff
@@ -428,6 +433,26 @@ TYPED_TEST(TestStringKernels, StrptimeDoesNotProvideDefaultOptions) {
   ASSERT_RAISES(Invalid, CallFunction("strptime", {input}));
 }
 
+TYPED_TEST(TestStringKernels, Join) {
+  auto separator = this->scalar("--");
+  CheckScalarBinary(
+      "binary_join", list(this->type()),
+      R"([["a", "bb", "ccc"], [], null, ["dd"], ["eee", null], ["ff", ""]])", separator,
+      this->type(), R"(["a--bb--ccc", "", null, "dd", null, "ff--"])");
```

Comment:
Another option could also be to skip nulls from the list, instead of skipping the list entirely if there is a null present? 
(so eg `["eee", null]` could give `"eee"` instead of ``null``? 

(don't really know what is most sensible to do, but generally for aggregations (and you can see a single join as a kind of aggregation?) nulls get skipped instead of propagated)

Reply:
My reasoning was that we can get that behavior using `fill_null`, while the current behavior would otherwise be difficult to get (we don't have any kernel that will replace list value with null if any of its list elements is null). I'm not sure if that's a good reason to have the current behavior though.

Reply:
Do you agree with my reasoning Joris? this current behavior is a bit more flexible, since with fill_null we can get the behavior you want.

Reply:
Sorry for the late reply here. Using `fill_null` (with empty string I suppose) would not be *fully* equivalent compared to skipping nulls, I think (assuming `["a", null, "b"]` with separator `"-"`, the result could be `"a--b"` vs `a-b`). 

That said, I am not sure myself what the best / most useful behaviour would be. So it's probably fine to pick whatever behaviour (not sure what's the easiest regarding implementation?) and potentially add an option for it later if there is demand for it (similar to how we are adding an option to the numeric reductions for skipping nulls or not).

Reply:
Looking at how pandas / R does this right now:

Pandas actually results in a missing value:

```
In [7]: s = pd.Series([["a", None, "b"], ["c", "d", "e"]])

In [8]: s.str.join("-")
Out[8]: 
0      NaN
1    c-d-e
dtype: object
```

but I doubt that this is necessarily intentional (when it comes to missing values). The docstring basically says that if there any non-string element in a list, the result is NaN for that list. And I assume this is just because in that case the stdlib `"-".join(..)` which is used under the hood fails.

For R, I am not fully sure what's the most "typical" way to do this, but with the basic `paste`, you actually get the string representation of NA:

```
> paste("a", NA, "b", sep="-")
[1] "a-NA-b"
```

(but `paste` is more general in converting input to strings first). 
Using a more specific string function from `stringr`, it actually results in NA:

```
> str_c("a", "b", sep="-")
[1] "a-b"
> str_c("a", NA, "b", sep="-")
[1] NA
```

where the docstring explicitly mentions this (*"Like most other R functions, missing values are "infectious": whenever a missing value is combined with another string the result will always be missing"*). So that could be an argument for returning null. But since in Arrow the default is that nulls are *not* infectious, it could also be an argument to skip it ;)

cc @ianmcook

Reply:
> So it's probably fine to pick whatever behaviour (not sure what's the easiest regarding implementation?) and potentially add an option for it later if there is demand for it (similar to how we are adding an option to the numeric reductions for skipping nulls or not)

Yes, although you could say you'd want a drop_null to implement the `["a", null, "b"]` -> `"a-b"` behavior. So I still think the current behaviour gives the most flexibility (assuming the existence of a drop_null for lists)

Reply:
I'm ok with this behaviour. We can add an option later if we want to override it.

Reply:
Regarding the R bindings: we will need some additional things beyond the scope of this PR to implement useful translations for `paste()` and `str_c()`—most importantly, a way of combining two or more arrays row-wise into list arrays (because in most real-world usage the strings users want to combine won't be in a list array). So I think it's fine to save those considerations for later.